aws/auth/mutable_static_creds_provider.cc

```diff
@@ -15,137 +15,36 @@
-  // This ensures that consumers will either see a version mismatch or see the updating state change.
-  //
+  std::uint64_t next_version = version_ + 1;
+  
 
-  current_.updating_ = true;
-  current_.creds_.SetAWSAccessKeyId(akid);
-  current_.creds_.SetAWSSecretKey(sk);
-  current_.creds_.SetSessionToken(token);
-  current_.version_++;
-  current_.updating_ = false;
-}
-
-bool MutableStaticCredentialsProvider::try_optimistic_read(Aws::Auth::AWSCredentials& destination) {
-  //
-  // This is an attempt to do an optimistic read.  We assume that the contents of the
-  // credentials may change in while copying to the result.
-  //
-  // 1. To handle this we first check if an update is in progress, and if it is we bounce out
-  // and retry.
-  //
-  // 2. If the credential isn't being updated we go ahead and load the current version of
-  // the credential.
-  //
-  // 3. At this point we go ahead and trigger a copy of the credential to the result.  This
-  // is what we will return if our remaining checks indicate everything is still ok.
-  //
-  // 4. We check again to see if the credential has entered updating, if it has it's possible
-  // the version of the credential we copied was split between the two updates.  So we
-  // discard our copy, and try again.
-  //
-  // 5. Finally we make check to see that the version hasn't changed since we started.  This
-  // ensures that the credential didn't enter and exit updates between our update checks.
-  //
-  // If everything is ok we are safe to return the credential, while it may be a nanosecnds
-  // out date it should be fine.
-  //
-
-  if (current_.updating_) {
-    //
-    // The credentials are currently being updated.  It's not safe to read so
-    // spin while we wait for the update to complete.
-    //
-    update_step(debug_stats.update_before_load_, debug_stats.retried_);
-    return false;
-  }
-  std::uint64_t starting_version = current_.version_;
-
-  //
-  // Trigger a trivial copy to the result
-  //
-  destination = current_.creds_;
-
-  if (current_.updating_) {
-    //
-    // The credentials object is being updated and the update may have started while we were
-    // copying it. For safety we discard what we have and try again.
-    //
-    update_step(debug_stats.update_after_load_, debug_stats.retried_);
-    return false;
-  }
-
-  std::uint64_t ending_version = current_.version_;
-
-  if (starting_version != ending_version) {
-    //
-    // The version changed in between the start of the copy, and the end
-    // of the copy.  We can no longer trust that the resulting copy is
-    // correct.  So we discard what we have and try again.
-    //
-    update_step(debug_stats.version_mismatch_, debug_stats.retried_);
-    return false;
-  }
-  update_step(debug_stats.success_);
-  return true;
+  std::shared_ptr<VersionedCredentials> new_credentials = std::make_shared<VersionedCredentials>(next_version, akid, sk, token);
+  std::atomic_store(&creds_, new_credentials);
```

Comment:
would it be good to add a comment that this is updating the global credentials ?